go-fuzz/mutator.go

```diff
@@ -221,23 +210,19 @@ func (m *Mutator) mutate(data []byte, ro *ROData) []byte {
 				continue
 			}
 			pos := m.rand(len(res) - 1)
+			buf := res[pos:]
 			v := uint16(interesting16[m.rand(len(interesting16))])
-			if m.r.Bool() {
-				v = swap16(v)
-			}
-			*(*uint16)(unsafe.Pointer(&res[pos])) = v
+			m.randByteOrder().PutUint16(buf, v)
 		case 13:
 			// Replace an uint32 with an interesting value.
 			if len(res) < 4 {
 				iter--
 				continue
 			}
 			pos := m.rand(len(res) - 3)
+			buf := res[pos:]
 			v := uint32(interesting32[m.rand(len(interesting32))])
-			if m.r.Bool() {
-				v = swap32(v)
-			}
-			*(*uint32)(unsafe.Pointer(&res[pos])) = v
+			m.randByteOrder().PutUint32(buf, v)
```

Comment:
How does this affect performance? I have a memory that encoding/binary is slow.

Reply:
Negligible impact. encoding/binary is fast now: The compiler recognizes the load pattern and converts it to a single unaligned load, equivalent to the unsafe magic, at least on amd64, which is where (I imagine) the vast majority of all fuzzing occurs. And the unsafe variant can crash on architectures that don't support unaligned loads; encoding/binary will at least be correct and slow. :)

Reply:
To be precise: LittleEndian loads are converted to a single unaligned load. BigEndian loads are converted to a single unaligned load and a BSWAPQ.